codelabs/Utopia Wholesale Codelab/Advanced Features/Deploying to Google Cloud Run.md

```diff
@@ -0,0 +1,279 @@
+   - The PORT environment variable can override this if needed
+
+2. **Base URL Configuration**: Added `base_url` pointing to the Cloud Run deployment URL
+   ```json
+   "base_url": "https://multipaz-openid4vci-server-971523157550.us-central1.run.app"
+   ```
+   - **Important**: Replace this with your own Cloud Run deployment URL after deploying the server
+   - This ensures all redirects and generated URLs use the correct public URL
+   - Without this, the server would generate URLs with `localhost:8080`
+
+3. **System of Record URL**: Added `system_of_record_url` pointing to the Records server
+   ```json
+   "system_of_record_url": "https://multipaz-records-server-971523157550.us-central1.run.app"
+   ```
+   - **Important**: Replace this with your own Records server Cloud Run URL after deploying it
+   - This connects the OpenID4VCI server to the Records server
+   - The OpenID4VCI server uses this to retrieve identity data when issuing credentials
+
+### **Changes to `multipaz-records-server`**
+
+The following changes were made to `multipaz-records-server/src/main/resources/resources/default_configuration.json`:
+
+1. **Port Configuration**: Changed `server_port` from `8004` to `8080`
+   - Consistent with Cloud Run's expected port
+
+2. **Base URL Configuration**: Added `base_url` pointing to the Cloud Run deployment URL
+   ```json
+   "base_url": "https://multipaz-records-server-971523157550.us-central1.run.app"
+   ```
+   - **Important**: Replace this with your own Cloud Run deployment URL after deploying the server
+   - Ensures proper redirects when accessing the root URL
+
+3. **Admin Password**: Added fixed `admin_password` set to `"multipaz-records"`
+   - Previously, the server generated a random password on each startup
+   - This makes the password predictable and manageable for deployment
+
+---
+
+## **Deployment Steps**
+
+### **Step 1: Create Dockerfile**
+
+Create a Dockerfile for the OpenID4VCI server in the `multipaz-openid4vci-server` directory:
+
+```bash
+mkdir -p multipaz-openid4vci-server && cat > multipaz-openid4vci-server/Dockerfile <<'EOF'
+FROM eclipse-temurin:17-jre
+WORKDIR /app
+COPY app.jar app.jar
+# Cloud Run sets PORT environment variable; pass it to the server using -param format
+CMD ["sh","-c","java -jar app.jar -param server_port=${PORT:-8080}"]
+EOF
+```
+
+**Explanation:**
+- `FROM eclipse-temurin:17-jre`: Uses Java 17 JRE base image (matches the server's Java version requirement)
+- `WORKDIR /app`: Sets the working directory inside the container
+- `COPY app.jar app.jar`: Copies the fat JAR file into the container
+- `CMD`: Runs the JAR with the PORT environment variable passed as a parameter
+  - Note: Uses `-param server_port=` format (not `--server.port=`) because this is a Ktor server, not Spring Boot
+  - `${PORT:-8080}` uses Cloud Run's PORT environment variable, defaulting to 8080 if not set
+
+### **Step 2: Build the Fat JAR**
```

Comment:
hmm, pls mention that this should be done locally & they should upload the jar to google cloud console using upload button on the top right side.